### PR TITLE
`azurerm_kubernetes_cluster`: Support for BYO `kubelet_identity`

### LINKED ISSUE #11980: Support Bring your own kubelet MI
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description
Request to add support to assign kubelet MI at cluster creation as per microsoft [docs](https://docs.microsoft.com/en-us/azure/aks/use-managed-identity#bring-your-own-control-plane-mi) so that it is possible to set UAMI as a kubelet identity.  

At the moment azurerm supports only export of the [kubelet_identity attributes](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster#kubelet_identity) but it doesn't support setting it as an AKS cluster attribute. 

With azure cli this can be done as documented [here](https://docs.microsoft.com/en-us/azure/aks/use-managed-identity#create-a-cluster-using-kubelet-identity).

### New or Affected Resource(s)
azurerm_kubernetes_cluster

### Potential Terraform Configuration
```
resource "azurerm_kubernetes_cluster" "example" {
  name                    = "aksexamplewithprivatednszone1"
  location                = azurerm_resource_group.example.location
  resource_group_name     = azurerm_resource_group.example.name
  dns_prefix              = "aksexamplednsprefix1"
  private_cluster_enabled = true
  private_dns_zone_id     = azurerm_private_dns_zone.example.id

  ... rest of configuration omitted for brevity

  kubeletidentity": {
    "clientId": "<client-id>",
    "objectId": "<object-id>",
    "resourceId": "/subscriptions/<subscriptionid>/resourcegroups/resourcegroups/providers/Microsoft.ManagedIdentity/userAssignedIdentities/myKubeletIdentity"
  }
```

### References
* microsoft docs on bring your own identity for kubelet MI:
https://docs.microsoft.com/en-us/azure/aks/use-managed-identity#bring-your-own-control-plane-mi

* azurerm_kubelet_identity
https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster#kubelet_identity

### PR DESCRIPTION
**Pull request by aristosvo**
Fixes #11980. Feature in Preview.

## Register before testing:
```bash
az feature register --namespace Microsoft.ContainerService -n CustomKubeletIdentityPreview
# wait until registered:
watch az feature list -o table --query "\"[?contains(name, 'Microsoft.ContainerService/CustomKubeletIdentityPreview')].{Name:name,State:properties.state}\""
az provider register --namespace Microsoft.ContainerService
```

## Acceptance Tests run:
- `TestAccKubernetesCluster_userAssignedKubeletIdentity` (with BYO Kubelet Identity)
- `TestAccKubernetesCluster_userAssignedIdentity` (without BYO Kubelet Identity)
```
❯ make acctests SERVICE='containers' TESTARGS='-run=TestAccKubernetesCluster_userAssigned'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./azurerm/internal/services/containers -run=TestAccKubernetesCluster_userAssigned -timeout 180m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
2021/06/02 13:43:06 [DEBUG] not using binary driver name, it's no longer needed
2021/06/02 13:43:08 [DEBUG] not using binary driver name, it's no longer needed
=== RUN   TestAccKubernetesCluster_userAssignedIdentity
=== PAUSE TestAccKubernetesCluster_userAssignedIdentity
=== RUN   TestAccKubernetesCluster_userAssignedKubeletIdentity
=== PAUSE TestAccKubernetesCluster_userAssignedKubeletIdentity
=== CONT  TestAccKubernetesCluster_userAssignedIdentity
=== CONT  TestAccKubernetesCluster_userAssignedKubeletIdentity
--- PASS: TestAccKubernetesCluster_userAssignedIdentity (615.75s)
--- PASS: TestAccKubernetesCluster_userAssignedKubeletIdentity (710.19s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/containers    714.228s
```

## ToDo:
- [x] Docs

### CODE REVIEW

**Review by mbfrahry — CHANGES_REQUESTED** [review]
This is looking good but I do have a doc question to make things a little more clear

File: website/docs/r/kubernetes_cluster.html.markdown [hunk 1]
```diff
@@ -366,6 +368,17 @@ An `identity` block supports the following:
 
 ---
 
+The `kubelet_identity` block supports the following:
+
+* `client_id` - (Optional) The Client ID of the user-defined Managed Identity to be assigned to the Kubelets. If not specified a Managed Identity is created automatically. Requires `kubelet_identity.0.object_id`,`kubelet_identity.0.user_assigned_identity_id` and `identity.0.user_assigned_identity_id` to be specified.
```
[the remark this answers is not in the document]
Comment (aristosvo) [reply website/docs/r/kubernetes_cluster.html.markdown]: Sounds like a plan! I moved the requirement on `identity.0.user_assigned_identity_id` to the generic note below.

Thanks!

**Review by mbfrahry — APPROVED** [review]
LGTM! Thanks @aristosvo!

**Comment by dwrusse** [comment]
Is there any way to enable the kubeletpreview provider via terraform?  I was hoping to embed this into a module that will be used on several different accounts.

**Comment by aristosvo** [comment]
@dwrusse Would https://www.terraform.io/docs/language/resources/provisioners/local-exec.html maybe work for you combined with `az cli`?

**Comment by dwrusse** [comment]
Potentially, yes.  I was hoping for a more "native" terraform method, but I agree with you that there likely isn't.  Thanks!

**Pull request merged**